docs/design/Impact Model Plugin.md

```diff
@@ -0,0 +1,195 @@
+
+#### Raises
+
+* `MissingConfig`
+* `IncorrectConfig`
+* `ExternalError`
+
+### Authentication
+
+
+#### Signature
+
+```ts
+public authenticate(authParams: AuthParams): void
+```
+
+#### Example usage
+
+```ts
+class ConcreteVM extends ImpactModelPluginInterface { ... }
+let model = new ConcreteVM().configure("backend-server", {vendor: "GCP"});
+try {
+    model.authenticate({username: "XXXX", password: "YYYY"})
+} catch AuthenticationError e {
+    ...
+}
+```
+
+#### Responsibilities
+
+* This is an optional call to authenticate using the model.
+* Some carbon models might be provided to users only through a commercial agreement or through an NDA. 
+* To restrict access to that, we need to have an authentication mechanism.
+* The spec does not currently prescribe a particular authentication mechanism, so it accepts multiple parameters to support multiple authentication methods.
+
+#### Parameters
+
+* **authParams** This is a dictionary of AuthenticationParams. Exact params depend on this model's authentication mechanism but could be a simple username/password, an API key, or another method.
+
+#### Returns
+
+Nothing. 
+
+#### Raises
+
+* `AuthenticationError`
+* `ExternalError`
+
+### Calculate
+
+This function estimates the emissions based on a single input observation. For each input observation, we calculate one output Impact Metric. Observations and Impact Metrics have a 1-1 mapping and relationship.
+
+#### Signature
+
+```ts
+public calculate(observations: Array<Observation>): Arrray<ImpactMetric>
+```
+
+#### Example usage
+
+```ts
+class ConcreteVM extends ImpactModelInterface { ... }
+let model = new ConcreteVM().configure("backend-server", {vendor: "GCP"});
+try {
+    let observation = {“date-time”: xxxx, “duration”: xxx, “cpu-util”: 0.5};
+    let impacts = model.calculate([observation]);
+    console.log(impacts);
+} catch {
+    ...
+}
+```
+
+#### Responsibilities
+
+* Checked that the passed in observations have all the required fields for this model.
+* Performs what validations it can that the provided observations are not malformed.
+* Passes the observations to the underlying carbon model, executes the model, and translates the response to match the emissions data type.
+
+#### Parameters
+
+* **observations** 
+  * This is an array of observations. 
+  * Each model can work with different types of inputs; the spec cannot predict what kinds of inputs will be used by all models, so we need to keep this very open. 
+  * It's an array since we will (in the future) need to deal with GridEMissions (`I`), and that requires input data in a fine grain to make sure we map to `I` at the same granularity. E.g., we might want to output carbon per hour, but the input energy data is in 5min increments so that we can make the grid emissions in the same 5 min increments.
+
+#### Returns
+
+An array of instances of an `ImpactMetric` holds the estimate of emissions.
+
+#### Raises
+
+* `NotAuthorized`
```

Comment:
JSII will not permit specific typed Exceptions. It can only throw generalised Error.

Reply:
Understood.